From bcb1a8a25e4ee9a94478378710de53b45a9b1517 Mon Sep 17 00:00:00 2001
From: Yuri Pankov <yuri.pankov@nexenta.com>
Date: Wed, 6 Dec 2017 08:19:31 +0300
Subject: [PATCH] OpenZFS 8898 - creating fs with checksum=skein on the boot
 pools fails ungracefully
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Authored by: Yuri Pankov <yuri.pankov@nexenta.com>
Reviewed by: Toomas Soome <tsoome@me.com>
Reviewed by: Andy Stormont <astormont@racktopsystems.com>
Reviewed-by: George Melikov <mail@gmelikov.ru>
Approved by: Dan McDonald <danmcd@joyent.com>
Ported-by: Brian Behlendorf <behlendorf1@llnl.gov>

OpenZFS-issue: https://www.illumos.org/issues/8898
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/9fa2266d9a
Closes #7031
---

Index: zfs/lib/libzfs/libzfs_dataset.c
===================================================================
--- zfs.orig/lib/libzfs/libzfs_dataset.c
+++ zfs/lib/libzfs/libzfs_dataset.c
@@ -27,7 +27,7 @@
  * Copyright (c) 2012 Pawel Jakub Dawidek <pawel@dawidek.net>.
  * Copyright (c) 2013 Martin Matuska. All rights reserved.
  * Copyright (c) 2013 Steven Hartland. All rights reserved.
- * Copyright 2016 Nexenta Systems, Inc.
+ * Copyright 2017 Nexenta Systems, Inc.
  * Copyright 2016 Igor Kozhukhov <ikozhukhov@gmail.com>
  * Copyright 2017 RackTop Systems.
  */
@@ -3507,6 +3507,11 @@ zfs_create(libzfs_handle_t *hdl, const c
 			    "pool must be upgraded to set this "
 			    "property or value"));
 			return (zfs_error(hdl, EZFS_BADVERSION, errbuf));
+
+		case ERANGE:
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "invalid property value(s) specified"));
+			return (zfs_error(hdl, EZFS_BADPROP, errbuf));
 #ifdef _ILP32
 		case EOVERFLOW:
 			/*
Index: zfs/man/man8/zfs.8
===================================================================
--- zfs.orig/man/man8/zfs.8
+++ zfs/man/man8/zfs.8
@@ -1057,6 +1057,10 @@ Please see
 for more information on these algorithms.
 .Pp
 Changing this property affects only newly-written data.
++.Pp
++Salted checksum algorithms
++.Pq Cm edonr , skein
++are currently not supported for any filesystem on the boot pools.
 .It Xo
 .Sy compression Ns = Ns Sy on Ns | Ns Sy off Ns | Ns Sy gzip Ns | Ns
 .Sy gzip- Ns Em N Ns | Ns Sy lz4 Ns | Ns Sy lzjb Ns | Ns Sy zle
